### PR TITLE
User service

### PR DESCRIPTION
**Pull request by GytisMas**

### CODE REVIEW

File: src/CSI.IBTA.DB.Migrations/Migrations/20240213111647_users-table.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+#nullable disable
+
+namespace CSI.IBTA.DB.Migrations.Migrations
+{
+    /// <inheritdoc />
+    public partial class userstable : Migration
```
Comment (radislavkireev) [inline src/CSI.IBTA.DB.Migrations/Migrations/20240213111647_users-table.cs]:
```suggestion
    public partial class Users: Migration
```

File: src/CSI.IBTA.Shared/DTOs/UserDto.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using CSI.IBTA.Shared.Entities;
+
+namespace CSI.IBTA.Shared.DTOs
+{
+    public record UserDto(int Id, string UserName, string FirstName, string LastName, int AccoundId, int EmployerId);
```
Comment (radislavkireev) [inline src/CSI.IBTA.Shared/DTOs/UserDto.cs:5]: Lower case the param names

File: src/CSI.IBTA.Shared/Settings/ImagesSettings.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿
+namespace CSI.IBTA.Shared.Settings
+{
+    public static class ImagesSettings
+    {
+        public static string[] GetAllowedFormats()
```
Comment (radislavkireev) [inline src/CSI.IBTA.Shared/Settings/ImagesSettings.cs]: Suggestion is to extract file formats to the settings file, so the formats are configurable. You may do it in a separate PR,  this one could be approved when you done with other notes.

File: src/CSI.IBTA.Shared/Utils/PasswordHasher.cs [hunk 4]
```diff
@@ -1,19 +1,18 @@
-﻿using CSI.IBTA.AuthService.Interfaces;
-using System.Security.Cryptography;
+﻿using System.Security.Cryptography;
 
-namespace CSI.IBTA.AuthService.Authentication
+namespace CSI.IBTA.Shared.Utils
 {
-    internal class PasswordHasher : IPasswordHasher
+    public static class PasswordHasher
```
Comment (radislavkireev) [inline src/CSI.IBTA.Shared/Utils/PasswordHasher.cs:5]: Lets keep it as is, but for the future  - lest avoid static classes, until they are extensions or super simple. 
They could not be mocked for the unit test which is disadvantage.

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+        public UsersService( 
+            IUnitOfWork unitOfWork)
+        {
+            _unitOfWork = unitOfWork;
+        }
+
+        public async Task<GenericResponse<UserDto[]>> GetAllUsers(HttpContext httpContext)
+        {
+            await _unitOfWork.Accounts.All();
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs]: EF should solve the mappings for you if you had map created, relations should not be solved manually.
Check this article: https://learn.microsoft.com/en-us/ef/core/modeling/relationships

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+            //.Select(user => new UserDto(user.Id, user.Account.Username, 
+            //user.Firstname, user.Lastname, user.Account.Id, employerId));
+            
+            return new GenericResponse<UserDto[]>(false, null, userList);
+        }
+
+        public async Task<GenericResponse<UserDto>> GetUser(int accountId, HttpContext httpContext)
+        {
+            await _unitOfWork.Accounts.All();
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs]: The same as above - relations should be handled by EF mappings, not manually. 
Also, in order to get 1 user, your code queries all Users + all employers + all relations between then. 
If we gonna have 10k users more that would be very slow and expensive operation.

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+                Account = newUser,
+            };
+
+            EmployerUser newEmployerUser = new EmployerUser()
+            {
+                Employer = employer.First(),
+                User = newUser
+            };
+
+            await _unitOfWork.Accounts.Add(newAccount);
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs]: User EF mappings, you should not manually solve relations. 
This note relates to the rest of the methods, I am not gonna repeat it, but all the methods of this class should be refactored. 
Ask colleagues to help and parallel the efforts if needed.

File: src/CSI.IBTA.UserService/Utils/FileUtils.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+﻿using CSI.IBTA.Shared.Settings;
+
+namespace CSI.IBTA.UserService.Utils
+{
+    internal static class FileUtils
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Utils/FileUtils.cs]: Extract i-face, do the regular class and inject it with DI.

**Review by radislavkireev — CHANGES_REQUESTED** [review]

File: src/CSI.IBTA.DB.Migrations/Migrations/20240214164641_Settings.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        /// <inheritdoc />
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "Settings",
+                columns: table => new
+                {
+                    Id = table.Column<int>(type: "int", nullable: false)
+                        .Annotation("SqlServer:Identity", "1, 1"),
+                    Condition = table.Column<string>(type: "nvarchar(max)", nullable: false),
```
Comment (mykytatishkin) [inline src/CSI.IBTA.DB.Migrations/Migrations/20240214164641_Settings.cs:19]: Why `nvarchar(max)`? nvarchar(50) or (100) isnt enough?

**Review by ttishkin — APPROVED** [review]

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        private readonly IUnitOfWork _unitOfWork;
+
+        public UsersService( 
+            IUnitOfWork unitOfWork)
+        {
+            _unitOfWork = unitOfWork;
+        }
+
+        public async Task<GenericResponse<UserDto[]>> GetAllUsers(HttpContext httpContext)
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs]: Do not pass httpContext to the business service. Business layer should not have a dependency on Web context

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+                .Select(user => new UserDto(user.Id, user.Account.Username, 
+                user.Firstname, user.Lastname, user.Account.Id,
+                user.Employer == null ? null : user.Employer.Id))
+                .ToArrayAsync();
+            
+            return new GenericResponse<UserDto[]>(false, null, userList);
+        }
+
+        public async Task<GenericResponse<UserDto>> GetUser(int accountId, HttpContext httpContext)
+        {
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs:36]: do not pass httpContext, but IPrincipal ( httpContext.User)

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+            return new GenericResponse<UserDto>(false, null,
+                new UserDto(user.Id, user.Account.Username, user.Firstname, user.Lastname, 
+                user.Account.Id, user.Employer == null ? -1 : user.Employer.Id)
+                );
+        }
+
+        public async Task<GenericResponse<NewUserDto>> CreateUser(CreateUserDto createUserDto, HttpContext httpContext)
+        {
+            if (!IsNextSuperiorRole(httpContext.User, createUserDto.Role))
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs]: do not pass httpContext, but IPrincipal ( httpContext.User)

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+                .Include(u => u.Emails)
+                .Include(u => u.Phones)
+                .FirstOrDefaultAsync(a => a.Id == userId);
+
+            if (user == null)
+            {
+                return new GenericResponse<UpdatedUserDto>(true, new HttpError("User not found", HttpStatusCode.NotFound), null);
+            }
+
+            var authUserId = httpContext.User.FindFirstValue(ClaimTypes.NameIdentifier);
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs]: do not pass httpContext, but IPrincipal ( httpContext.User)

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+                new UpdatedUserDto(user.Id, user.Account.Username, user.Account.Password
+                , user.Firstname, user.Lastname, user.Account.Id, user.Account.Role
+                , user.Phones[0].PhoneNumber, user.Emails[0].EmailAddress
+                , user.Addresses[0].State, user.Addresses[0].Street, user.Addresses[0].City, user.Addresses[0].Zip)
+                );
+        }
+
+        public async Task<GenericResponse<bool>> DeleteUser(int userId, HttpContext httpContext)
+        {
+            var user = await _unitOfWork.Users
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs:185]: do not pass httpContext, it is not even used here

File: src/CSI.IBTA.UserService/Services/UsersService.cs [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+using CSI.IBTA.Shared.Entities;
+using CSI.IBTA.Shared.Utils;
+using CSI.IBTA.UserService.Interfaces;
+using System.Net;
+using System.Security.Claims;
+using Microsoft.EntityFrameworkCore;
+
+namespace CSI.IBTA.UserService.Services
+{
+    internal class UsersService : IUsersService
```
Comment (radislavkireev) [inline src/CSI.IBTA.UserService/Services/UsersService.cs:12]: You don't need to verify httpCOntext.User permissions at UsersService if it is already done at UserController. 
So, my vote would be to verify roles and credentials in controllers

**Review by radislavkireev — CHANGES_REQUESTED** [review]

**Review by radislavkireev — APPROVED** [review]

**Pull request merged**